search/search_quality/scoring_model.py

```diff
@@ -64,121 +79,62 @@ def compute_ndcg_for_w(data, w):
-        for i in range(n):
-            w[i] += self.eps
-            g[i] = (self.value(w) - fw) / self.eps
-            w[i] -= self.eps
-        return g
-
-
-class RankingSVMLoss:
-    """
-    Represents a loss function with a gradient for a RankingSVM model.
-    Simple version of a loss function for a ranked list of features
-    has following form:
-
-    loss(w) = sum{i < j: max(0, 1 - sign(y[j] - y[i]) * dot(w, x[j] - x[i]))} + lam * dot(w, w)
-
-    This version is slightly modified, as we are dealing with a group
-    of ranked lists, so loss function is actually a weighted sum of
-    loss values for each list, where each weight is a 1 / list size.
-    """
-
-    @staticmethod
-    def sign(x):
-        if x < 0:
-            return -1
-        elif x > 0:
-            return 1
-        return 0
+    grouped = data.groupby(data['SampleId'], sort=False).groups
 
+    xs, ys = [], []
+    for id in grouped:
+        indices = grouped[id]
+        features = data.ix[indices][FEATURES]
+        relevances = np.array(data.ix[indices]['Relevance'])
 
-    def __init__(self, data, lam=1e-3):
-        self.coeffs, self.lam = [], lam
+        n, total = len(indices), 0
+        for k, (i, j) in enumerate(itertools.combinations(range(n), 2)):
+            y = np.sign(relevances[j] - relevances[i])
+            if y == 0:
+                continue
+            x = (np.array(features.iloc[j]) - np.array(features.iloc[i]))
+            xs.append(x)
+            ys.append(y)
+            total = total + 1
 
-        grouped = data.groupby(data['SampleId'], sort=False).groups
-        for id in grouped:
-            indices = grouped[id]
-            features = data.ix[indices][FEATURES]
-            relevances = np.array(data.ix[indices]['Relevance'])
-            n = len(indices)
-            for i in range(n):
-                for j in range(i + 1, n):
-                    y = self.sign(relevances[j] - relevances[i]) / n
-                    dx = y * (np.array(features.iloc[j]) - np.array(features.iloc[i]))
-                    self.coeffs.append(dx)
+        # Scales this group of features to equalize different search
+        # queries.
+        for i in range(-1, -total, -1):
+            xs[i] = xs[i] / total
+    return xs, ys
 
 
-    def value(self, w):
-        result = self.lam * np.dot(w, w)
-        for coeff in self.coeffs:
-            v = 1 - np.dot(coeff, w)
-            if v > 0:
-                result += v
-        return result
+def main(args):
+    data = pd.read_csv(sys.stdin)
+    normalize_data(data)
+    x, y = transform_data(data)
 
+    clf = svm.LinearSVC(random_state=args.seed)
+    cv = cross_validation.KFold(len(y), n_folds=5, shuffle=True, random_state=args.seed)
 
-    def gradient(self, w):
-        result = 2 * self.lam * w
-        for coeff in self.coeffs:
-            if 1 - np.dot(coeff, w) > 0:
-                result = result - coeff
-        return result
+    # Let's find regularizer by a grid search.
```

Comment:
It's not obvious that C is a regularizer constant. But if you insist, I'll remove the comment.